rfcs/text/0014_notification_service.md

```diff
@@ -0,0 +1,373 @@
+- *Sourcing* - collects notification events created by different internal sources.
+- *Notification* system - abstracts the way notifications are created, stored, configured, and retrieved.  Ensures compatible operation for on-prem and Cloud Kibana versions.
+- *Delivery* - provides API for Kibana UI to retrieve the latest notifications for a user, to update notification status (read / unread).
+- *Settings* - provides API to adjust user-specific notification settings (unsubscribe from / subscribe to a source).
+![image](../images/0014/kns_key_elements.png)
+
+### Notification event
+It consists of a message with an associated CTA to show to an end-user.
+*NotificationEvent* interface
+- `recipient_id: string []` - unique identifiers of all the recipients. It might be the same as `elastic_id` (if the env supports it) - the unique user identifier across all the company products. 
+There is currently no way to unambiguously identify the user. WIP [#82725](https://github.com/elastic/kibana/issues/82725)
+- `priority: number` - to prioritize the delivery in case of async delivery mechanism.
+Where 0 - the standard priority, 10 - the critical priority. An event with high priority is processed first.
+Shouldn’t be used in UI.
+- `created_at: number` - Unix timestamp in UTC timezone when a notification has been created.
+- `expired_at?: number` - Unix timestamp in UTC timezone when notification can be removed from the system. 
+- `pinned_status?: boolean` - a flag to pin a notification at top of the list of notifications. Used by a source to draw a user's attention to a notification.
+- `source_type: string` - source type. the same as source domain name: `cloud`, `alerting`.
+- `source_subtype: string` - source-specific type (`cloud.insufficient_funds`, `alerting.something`)
+- `group_id?: string` - identifier to associate several notifications in the UI
+- `content: NotificationContent` - see below.
+
+*NotificationContent* interface (see [UI component](https://github.com/elastic/eui/issues/4257))
+- `icon?: { euiIconType: string }` - icon on the left side of the title.
+Compatible with all the types supported by euiIcon type: a predefined EUI icon name, URL to SVG file, data URL for SVG file.
+- `badge?: "critical"` - badge below the notification message.
+- `message: { i18n_key: string, values: object } | { text: string }` - a notification message is subject to i18n. 
+Locale might be changed on the Space level (on the User level in the future).
+Thus, we cannot determine a message locale during notification creation.
+Considering that Kibana notifications might be rendered outside of Kibana UI (in Cloud UI, for example), the message must contain translations for all the supported locales.
+- `title: { i18n_key: string, values: object, url?: string } | { text: string, url?: string }` - notification header.
+- `action: { i18n_key: string, values: object, url: string } | { text: string, url: string }` - notification CTA
```

Comment:
> I'm still concerned about i18n keys changing between releases. We don't have enough rigor or tooling in place to be sure that the keys don't change and then break notifications created before the change (or sourced from different Stack versions). 

That's a valid concern.  We are going to store **translations** for a given notification:
```
Considering that Kibana notifications might be rendered outside of Kibana UI (in Cloud UI, for example), the message must contain translations for all the supported locales.
```
we need `i18n_key` to retrieve all the translations from `@kbn/i18n` before storing them in an index or sending them to UNS.

Reply:
Could we simplify things by only storing the translation for the users current language preference? Old notifications won't change when a user changes their language prefs, but that seems like an acceptable tradeoff.

Reply:
A reasonable solution, I will update RFC 👍

Reply:
> Could we simplify things by only storing the translation for the users current language preference? Old notifications won't change when a user changes their language prefs, but that seems like an acceptable tradeoff.

@alexfrancoeur How do you feel about this from a product point of view? It's not the best UX, but might be an acceptable trade-off.

Reply:
I talked about this with @alexfrancoeur a bit yesterday. In general, I don't like when surprising things happen in our product and having UI be in multiple languages depending on when the state of a cluster setting changes is surprising in my opinion. That being said, language setting changes are probably infrequent and it is my understanding that our internationalization sophistication is still maturing today, so we agreed this would be okay.

If we put ourselves in a user's shoes, the worst case scenario is probably something like: English is the default language for a deployment, I bring a deployment up, the deployment experiences problems and generates notifications about what went wrong, I do not read English, I switch my deployment language expecting to be able to return and read those messages, I am frustrated that this expectation was not met.

It will be a while before the notification system and internationalization is capable of producing an experience like that, but it's good to paint a real picture.

In case we want to look back at this for context, are our concerns around storing multiple languages in a notification driven by disk concerns or network concerns?

Reply:
> In case we want to look back at this for context, are our concerns around storing multiple languages in a notification driven by disk concerns or network concerns?

mostly yes.

> I agree that it seems simpler to store the translated text. Spaces does throw an interesting wrench in that plan though... Perhaps if the source of the notification event is some space-specific entity, we use the Space's locale if the user hasn't specified one?

@kobelb  That will require extending User profile API to retrieve user-specific settings (once we have them). Worth noting that Kibana doesn't have access to user credentials at notification creation time. Probably we can mark some settings as `public` not to apply the security model to them. 

In general, this approach might not give us a lot of benefits if several users set different locales because we have to duplicate a notification for every locale. 
For example, Alert is triggered for UserA, UserB, UserC:
UserA and UserB have locale `en`: we store a notification with a message in English.
UserC has locale `de`: we store the same notification with the message in German.

So instead of storing the single notification object with `en` and `de` locales, we persist two notification objects, which ends up consuming even more disk space.

Reply:
@jowiho @kobelb What do you think?

Considering that at the current moment there is no way to get user locale without user credentials, I'd suggest either starting with storing all possible translations for a message or storing translations for `en` locale only.

Reply:
@restrry as we discussed over Zoom, I think that both options are fine for the short-term. Long-term, I think that we're going to end up having to store all possible translations to provide the ideal UX.